init.lua

```diff
@@ -628,32 +628,54 @@ require('lazy').setup({
 
       -- Diagnostic Config
       -- See :help vim.diagnostic.Opts
-      vim.diagnostic.config {
-        severity_sort = true,
-        float = { border = 'rounded', source = 'if_many' },
-        underline = { severity = vim.diagnostic.severity.ERROR },
-        signs = vim.g.have_nerd_font and {
-          text = {
-            [vim.diagnostic.severity.ERROR] = '󰅚 ',
-            [vim.diagnostic.severity.WARN] = '󰀪 ',
-            [vim.diagnostic.severity.INFO] = '󰋽 ',
-            [vim.diagnostic.severity.HINT] = '󰌶 ',
-          },
-        } or {},
-        virtual_text = {
-          source = 'if_many',
-          spacing = 2,
-          format = function(diagnostic)
-            local diagnostic_message = {
-              [vim.diagnostic.severity.ERROR] = diagnostic.message,
-              [vim.diagnostic.severity.WARN] = diagnostic.message,
-              [vim.diagnostic.severity.INFO] = diagnostic.message,
-              [vim.diagnostic.severity.HINT] = diagnostic.message,
-            }
-            return diagnostic_message[diagnostic.severity]
-          end,
-        },
-      }
+      -- Diagnostic configuration similar to VS Code's Error Lens.
+      -- In insert mode, diagnostics are displayed as inline virtual text.
+      -- In normal mode, diagnostics are shown as virtual lines below the affected lines.
+
+      local function set_virtual_text(enable)
+        local diagnostic_icons = {
+          [vim.diagnostic.severity.ERROR] = '󰅚 ',
+          [vim.diagnostic.severity.WARN] = '󰀪 ',
+          [vim.diagnostic.severity.INFO] = '󰋽 ',
+          [vim.diagnostic.severity.HINT] = '󰌶 ',
+        }
+
+        vim.diagnostic.config {
+          update_in_insert = true, -- error messages in insert mode
+          severity_sort = true,
+          float = { border = 'rounded', source = 'if_many' },
+          underline = { severity = vim.diagnostic.severity.ERROR },
+          signs = vim.g.have_nerd_font and {
+            text = diagnostic_icons,
+          } or {},
+
+          virtual_lines = not enable and {
+            format = function(diagnostic)
+              return (diagnostic_icons[diagnostic.severity] or '') .. diagnostic.message
+            end,
+          } or false,
+
+          virtual_text = enable and {
+            source = 'if_many',
+            spacing = 2,
+            format = function(diagnostic)
+              return (diagnostic_icons[diagnostic.severity] or '') .. diagnostic.message
+            end,
+          } or false,
+        }
+      end
```

Comment:
I believe the logic could be simplified by settings most of the config outside of the function.
i.e. something like this:
```suggestion
      vim.diagnostic.config {
        severity_sort = true,
        float = { border = 'rounded', source = 'if_many' },
        underline = { severity = vim.diagnostic.severity.ERROR },
        signs = vim.g.have_nerd_font and {
          text = {
            [vim.diagnostic.severity.ERROR] = '󰅚 ',
            [vim.diagnostic.severity.WARN] = '󰀪 ',
            [vim.diagnostic.severity.INFO] = '󰋽 ',
            [vim.diagnostic.severity.HINT] = '󰌶 ',
          },
        } or {},
      }

      -- Diagnostic configuration similar to VS Code's Error Lens.
      -- In insert mode, diagnostics are displayed as inline virtual text.
      -- In normal mode, diagnostics are shown as virtual lines below the affected lines.
      ---@param enable boolean
      local function set_virtual_text(enable)
        vim.diagnostic.config {
          virtual_lines = not enable,

          virtual_text = enable and {
            source = 'if_many',
            spacing = 2,
          },
        }
      end
```

Reply:
Also in my example, I removed `format` as suggested by https://github.com/nvim-lua/kickstart.nvim/pull/1550

Reply:
Thank you for the feedback and suggestions! 🙌

I've updated the code to reflect the recommendations:

- I've kept the severity icons for both `virtual_text` and `virtual_lines`, as I believe it improves the visual clarity and overall diagnostic experience — especially when switching between modes.
- I've added a call to `set_virtual_text(false)` right after defining the function. While I didn’t observe any issues during testing without it, I agree it's a good addition for consistency and to prevent potential edge cases.
- I also set `underline = true` instead of restricting it to only errors. I found that underlining all severities helps quickly locate the affected lines, improving usability during review and debugging.

If there's still interest in merging this feature, I’d be happy to re-submit the PR with these improvements applied.

Thanks again for the thoughtful review!

```lua


 -- Diagnostic Config
      -- See :help vim.diagnostic.Opts
      local diagnostic_icons = {
        [vim.diagnostic.severity.ERROR] = '󰅚 ',
        [vim.diagnostic.severity.WARN] = '󰀪 ',
        [vim.diagnostic.severity.INFO] = '󰋽 ',
        [vim.diagnostic.severity.HINT] = '󰌶 ',
      }

      vim.diagnostic.config {
        update_in_insert = true,
        severity_sort = true,
        float = { border = 'rounded', source = 'if_many' },
        underline = true, -- { severity = vim.diagnostic.severity.ERROR }
        signs = vim.g.have_nerd_font and { text = diagnostic_icons } or {},
      }

      -- Diagnostic configuration similar to VS Code's Error Lens.
      -- In insert mode, diagnostics are displayed as inline virtual text.
      -- In normal mode, diagnostics are shown as virtual lines below the affected lines.
      ---@param enable boolean
      local function set_virtual_text(enable)
        vim.diagnostic.config {
          virtual_lines = not enable and {
            format = function(diagnostic)
              return (diagnostic_icons[diagnostic.severity] or '') .. diagnostic.message
            end,
          } or false,
          virtual_text = enable and {
            source = 'if_many',
            spacing = 2,
            format = function(diagnostic)
              return (diagnostic_icons[diagnostic.severity] or '') .. diagnostic.message
            end,
          } or false,
        }
      end

      set_virtual_text(false)

      vim.api.nvim_create_autocmd('InsertEnter', {
        callback = function()
          set_virtual_text(true)
        end,
      })

      vim.api.nvim_create_autocmd('InsertLeave', {
        callback = function()
          set_virtual_text(false)
        end,
      })




```

Reply:
> If there's still interest in merging this feature, I’d be happy to re-submit the PR with these improvements applied.

Go for it. 
I think it would be a great addition for kickstart (or at least for my config, if the maintainers don't agree with me :)